source/core_mqtt.c

```diff
@@ -1269,22 +1285,31 @@ static MQTTStatus_t sendPublishAcks( MQTTContext_t * pContext,
 
         if( status == MQTTSuccess )
         {
+            MQTT_PRE_SEND_HOOK( pContext );
+
             /* Here, we are not using the vector approach for efficiency. There is just one buffer
              * to be sent which can be achieved with a normal send call. */
             bytesSent = sendBuffer( pContext,
                                     localBuffer.pBuffer,
                                     MQTT_PUBLISH_ACK_PACKET_SIZE );
+
+            MQTT_POST_SEND_HOOK( pContext );
         }
 
         if( bytesSent == ( int32_t ) MQTT_PUBLISH_ACK_PACKET_SIZE )
         {
             pContext->controlPacketSent = true;
+
```

Comment:
Would it be desirable to keep hold of the mutex between the sendPacket and MQTT_UpdateStateAck calls?

Reply:
I think not.
They are decoupled altogether. Also, the state will be protected by its mutex.